wfe2/verify.go

```diff
@@ -18,233 +20,511 @@ import (
 	"gopkg.in/square/go-jose.v2"
 )
 
-// signatureValidationError indicates that the user's signature could not
-// be verified, either through adversarial activity, or misconfiguration of
-// the user client.
-type signatureValidationError string
-
-func (e signatureValidationError) Error() string { return string(e) }
-
-func algorithmForKey(key *jose.JSONWebKey) (string, error) {
-	switch k := key.Key.(type) {
-	case *rsa.PublicKey:
-		return string(jose.RS256), nil
-	case *ecdsa.PublicKey:
-		switch k.Params().Name {
-		case "P-256":
-			return string(jose.ES256), nil
-		case "P-384":
-			return string(jose.ES384), nil
-		case "P-521":
-			return string(jose.ES512), nil
-		}
+const sigAlgErr = "no signature algorithms suitable for given key type"
+
+func sigAlgorithmForECDSAKey(key *ecdsa.PublicKey) (jose.SignatureAlgorithm, error) {
+	params := key.Params()
+	switch params.Name {
+	case "P-256":
+		return jose.ES256, nil
+	case "P-384":
+		return jose.ES384, nil
+	case "P-521":
+		return jose.ES512, nil
 	}
-	return "", signatureValidationError("no signature algorithms suitable for given key type")
+	return "", fmt.Errorf(sigAlgErr)
+}
+
+func sigAlgorithmForKey(key interface{}) (jose.SignatureAlgorithm, error) {
+	switch k := key.(type) {
+	case *rsa.PublicKey, rsa.PublicKey, *rsa.PrivateKey, rsa.PrivateKey:
```

Comment:
> Let's take only the pointer version of these types, the same as the native Go library does

Sure, no problem.

> Similarly, is there a reason for this function to accept both public and private keys?

It was [originally written](https://github.com/letsencrypt/boulder/blob/2a84bc2495b79e7687ee5cde55e61d31d6371374/wfe/jose.go#L11:L26) for just public keys. I added support for private keys as well because it made it slightly easier to use when constructing `jose.SigningKey`'s in tests (e.g. [this instance](https://github.com/cpu/boulder/blob/8db7370b897be9a5a505b95d19e60ef30ed4d3be/wfe2/verify_test.go#L59)) when I have a private key. I'll remove this change and have the tests use the private key's public key instead.